### PR TITLE
Feature - 39520 - Playwright Support Existing POM

### PR DESCRIPTION
**Pull request by IamRanjeetSingh**
Thank you for your contribution.
Before submitting this PR, please make sure:

- [ ] PR description and commit message should describe the changes done in this PR
- [ ] Verify the PR is pointing to correct branch i.e. Release or Beta branch if the code fix is for specific release , else point it to master
- [ ] Latest Code from master or specific release branch is merged to your branch
- [ ] No unwanted\commented\junk code is included
- [ ] No new warning upon build solution
- [ ] Code Summary\Comments are added to my code which explains what my code is doing
- [ ] Existing unit test cases are passed
- [ ] New Unit tests are added for your development
- [ ] Sanity Tests are successfully executed for New and Existing Functionality
- [ ] Verify that changes are compatible with all relevant browsers and platforms.
- [ ] After creating pull request there should not be any conflicts
- [ ] Resolve all Codacy comments
- [ ] Builds and checks are passed before PR is sent for review
- [ ] Resolve code review comments
- [ ] Update the Help Library document to match any feature changes

## Summary by CodeRabbit

- **New Features**
  - Introduced screen-related operations and image handling specifically for Windows platforms.
  - Added new methods to capture and manipulate screen information and images, including merging images vertically.
  - Implemented functionality for taking and handling screenshots within web automation.

- **Refactor**
  - Updated namespaces and method structures for better readability and asynchronous operations.
  - Renamed methods to standardize naming conventions related to URL, title, and logs in browser tabs.

- **Enhancements**
  - Added new methods and interfaces to support asynchronous screenshot capturing and screen information retrieval.
  - Enhanced element locating and interaction logic in web driver modules.

### CODE REVIEW

**Review by Maheshkale447 — CHANGES_REQUESTED** [review]

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/ActionHandlers/ActUIElementHandler.cs [hunk 1]
```diff
@@ -32,180 +36,240 @@ internal sealed class ActUIElementHandler
...
+            string pathToElement = elementInfo.Path;
+            if (!elementInfo.IsAutoLearned)
+            {
+                GingerCore.ValueExpression valueExpression = new(_environment, _businessFlow);
+                pathToElement = valueExpression.Calculate(pathToElement);
+            }
+
+            await _browser.CurrentWindow.CurrentTab.SwitchToMainFrameAsync();
+            
+            if (string.IsNullOrEmpty(pathToElement))
```
[the remark this answers is not in the document]
Comment (IamRanjeetSingh) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/ActionHandlers/ActUIElementHandler.cs:216]: handled

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/ActionHandlers/ActUIElementHandler.cs [hunk 2]
```diff
@@ -32,180 +36,240 @@ internal sealed class ActUIElementHandler
...
+            foreach (string iframePath in iframesPaths)
+            {
+                await _browser.CurrentWindow.CurrentTab.SwitchFrameAsync(eLocateBy.ByRelXPath, iframePath);                
+            }
         }
 
         private async Task HandleClickOperationAsync()
         {
-            IBrowserElement element = await GetFirstMatchingElementAsync();
+            IBrowserElement element = await GetFirstMatchingElementAsync(_act.ElementLocateBy, _act.ElementLocateValueForDriver);
```
[the remark this answers is not in the document]
Comment (IamRanjeetSingh) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/ActionHandlers/ActUIElementHandler.cs:233]: handled

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/POMLocatorParser.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
+﻿using amdocs.ginger.GingerCoreNET;
```
[the remark this answers is not in the document]
Comment (IamRanjeetSingh) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/POMLocatorParser.cs:1]: done

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightDriver.cs [hunk 4]
```diff
@@ -126,7 +126,7 @@ public override void RunAction(Act act)
                     actBrowserElementHandler.HandleAsync().Wait();
                     break;
                 case ActUIElement actUIElement:
-                    ActUIElementHandler actUIElementHandler = new(actUIElement, _browser!);
+                    ActUIElementHandler actUIElementHandler = new(actUIElement, _browser!, BusinessFlow, Environment);
```
[the remark this answers is not in the document]
Comment (IamRanjeetSingh) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightDriver.cs:129]: this method is a sync and we cannot change it's method signature right now

File: Ginger/GingerCore/GeneralLib/General.cs [hunk 5]
```diff
@@ -1375,6 +1376,168 @@
...
+            switch(imageFormat)
+            {
+                case ImageFormat.Png:
+                    return System.Drawing.Imaging.ImageFormat.Png;
+                case ImageFormat.Jpeg:
+                    return System.Drawing.Imaging.ImageFormat.Jpeg;
+                default:
+                    throw new Exception($"Unknown ImageFormat '{imageFormat}'");
+            }
+        }
```
[the remark this answers is not in the document]
Comment (IamRanjeetSingh) [reply Ginger/GingerCore/GeneralLib/General.cs]: will add method summary in different PR

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/POMElementLocator.cs [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        private string EvaluteLocateValue(string locateValue)
+        {
+            GingerCore.ValueExpression valueExpression = new(_environment, _businessFlow);
+            return valueExpression.Calculate(locateValue);
+        }
+
+        private void UpdatePOM()
+        {
+            throw new NotImplementedException();
+        }
```
[the remark this answers is not in the document]
Comment (IamRanjeetSingh) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/POMElementLocator.cs:136]: will handle in next PR

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/POMElementLocator.cs [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+            {
+                elements = [];
+            }
+
+            return new LocateResult()
+            {
+                Elements = elements,
+                WasAutoUpdated = wasAutoUpdated,
+            };
+        }
```
[the remark this answers is not in the document]
Comment (IamRanjeetSingh) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/POMElementLocator.cs:82]: We want the existing implementation to retry.

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/POMElementLocator.cs [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+        {
+            GingerCore.ValueExpression valueExpression = new(_environment, _businessFlow);
+            return valueExpression.Calculate(locateValue);
+        }
+
+        private void UpdatePOM()
+        {
+            throw new NotImplementedException();
+        }
+    }
```
[the remark this answers is not in the document]
Comment (IamRanjeetSingh) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/POMElementLocator.cs:137]: Those classes are relevant for POMElementLocator, so need them as it is

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/ActionHandlers/ActUIElementHandler.cs [hunk 9]
```diff
@@ -32,180 +36,240 @@ internal sealed class ActUIElementHandler
...
+
+            await _browser.CurrentWindow.CurrentTab.SwitchToMainFrameAsync();
+            
+            if (string.IsNullOrEmpty(pathToElement))
+            {
+                return;
+            }
+
+            //split Path by commo outside of brackets 
+            Regex splitByCommaOutsideBrackets = new(@",(?![^\[]*[\]])");
```
[the remark this answers is not in the document]
Comment (IamRanjeetSingh) [reply Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/ActionHandlers/ActUIElementHandler.cs:222]: defined as static

**Review by Maheshkale447 — APPROVED** [review]

**Pull request merged**